File: scripts/decompile_mapping.py
# Get mapping of decompiler statements to lines of assembly
# @author N/A
# @category selection
# @keybinding Alt-d
# @menupath Select.Program Highlight.Decompiler
# @toolbar cache.png
from __future__ import unicode_literals

from ghidra.program.model.listing.CodeUnit import *
from ghidra.program.model.listing.Listing import *
from ghidra.app.decompiler import DecompInterface
from ghidra.app.decompiler.component import DecompilerUtils

import sys
import os
import json
import string
import traceback

printable = string.ascii_letters + string.digits + string.punctuation + " "


def log(msg):
    print("decompile_mapping.py: %s" % msg)


def hex_escape(s):
    return "".join(c if c in printable else r"\x{0:02x}".format(ord(c)) for c in s)


# get ghidra root directory
ghidra_default_dir = os.getcwd()

# get ghidra jython directory
jython_dir = os.path.join(
    ghidra_default_dir,
    "Ghidra",
    "Features",
    "Python",
    "lib",
    "Lib",
    "site-packages",
)

# insert jython directory into system path
sys.path.insert(0, jython_dir)


def build_mapping():
    log("build_mapping() starting")
    decomp = DecompInterface()
    decomp.openProgram(currentProgram)
    functionManager = currentProgram.getFunctionManager()
    function_count = functionManager.getFunctionCount()
    log("opened program %r with %d functions" % (currentProgram.getName(), function_count))

    output_map = {
        "meta": {
            "load_addr": str(currentProgram.getImageBase()),
            # optional: per-function metadata (not used by extract(), but handy)
            "functions": {},
            "function_count": function_count,
        }
    }

    processed_count = 0
    failed_count = 0

    for fun in functionManager.getFunctions(True):
        # Iterate in entry-point order
        funEntry = fun.getEntryPoint()
        function = functionManager.getFunctionContaining(funEntry)
        if function is None:
            continue

        # Use fully-qualified name to distinguish Lexer::getObj(int) from Parser::getObj(int)
        func_name = function.getName(True)

        # Set up structure for this function if needed
        if func_name not in output_map:
            output_map[func_name] = {}
            # store metadata under meta, NOT under the function dict
            output_map["meta"]["functions"][func_name] = {"entry": str(funEntry)}

        try:
            results = decomp.decompileFunction(function, 120, monitor)
            if not results or not results.decompileCompleted():
                log("decompilation did not complete for %s" % func_name)
                # skip functions that failed to decompile
                continue
            markup = results.getCCodeMarkup()
            if markup is None:
                log("no C markup returned for %s" % func_name)
                continue

            # Not strictly needed, but kept from original script
            highfun = markup.getHighFunction()
            clang = markup.getClangFunction()

            # Iterate over lines of source (add a stable line index for downstream ordering)
            for i, line in enumerate(DecompilerUtils.toLines(markup), start=1):
                unicode_line = hex_escape(line.toString())
                if len(unicode_line) > 30 and "WARNING:" in unicode_line[:30]:
                    continue

                # Prefix with an index so downstream can parse "NNN: <code>"
                tagged_line = "%d: %s" % (i, unicode_line)

                # Track whether we associated this source line with any address
                lineAdded = False

                # Tokens on this source line (may be missing depending on Ghidra/Jython object)
                tokens = getattr(line, "allTokens", None) or []

                for tok_in_line in tokens:
                    min_addr_obj = tok_in_line.getMinAddress()

                    # Some tokens in C do not correspond to an address such as '('
                    if min_addr_obj is None:
                        continue

                    minAddr = str(min_addr_obj)  # e.g., "0010c15a"

                    # Ensure per-address dict exists
                    entry = output_map[func_name].setdefault(minAddr, {})

                    # Add source line (avoid duplicates)
                    lines_list = entry.setdefault("line", [])
                    if tagged_line not in lines_list:
                        lines_list.append(tagged_line)

                    # Add token text (optionally dedupe if you want)
                    tok_str = hex_escape(tok_in_line.toString())
                    entry.setdefault("tokens", []).append(tok_str)

                    lineAdded = True

                # If no token on this line mapped to an address, store under "None"
                if not lineAdded:
                    entry = output_map[func_name].setdefault("None", {})
                    lines_list = entry.setdefault("line", [])
                    if tagged_line not in lines_list:
                        lines_list.append(tagged_line)

            processed_count += 1

        except Exception as e:
            # Log which function blew up, but keep going
            log("ERROR in function %s: %s" % (func_name, e))
            traceback.print_exc()
            failed_count += 1

            # Track failed functions in meta for debugging
            failed = output_map["meta"].get("failed_functions")
            if failed is None:
                output_map["meta"]["failed_functions"] = [func_name]
            elif func_name not in failed:
                failed.append(func_name)

            continue

    output_map["meta"]["processed_functions"] = processed_count
    output_map["meta"]["failed_function_count"] = failed_count
    log("build_mapping() finished: processed=%d failed=%d" % (processed_count, failed_count))
    return output_map


def main():
    log("script started")
    args = getScriptArgs()
    log("args = %r" % (args,))

    if not args:
        raise RuntimeError("No output path argument provided")

    out_path = args[0]
    out_dir = os.path.dirname(out_path)
    if out_dir and not os.path.isdir(out_dir):
        log("creating output directory %s" % out_dir)
        os.makedirs(out_dir)

    try:
        mapping = build_mapping()
    except Exception as e:
        # If *something* unexpected happens at the top level, still emit a file
        log("ERROR while building mapping: %s" % e)
        traceback.print_exc()
        mapping = {
            "meta": {
                "error": str(e),
                "exception_type": e.__class__.__name__,
            }
        }

    log("writing JSON to %s" % out_path)
    f = open(out_path, "w")
    try:
        # Some instructions do not map to anything, including anything in prologue of function
        json.dump(mapping, f)
    finally:
        f.close()
    log("successfully wrote JSON output")


# Run immediately when invoked by Ghidra/headless
try:
    main()
except Exception as e:
    log("FATAL top-level error: %s" % e)
    traceback.print_exc()
    raise

File: scripts/decompile_mapping.java
/*
Copyright 2023 National Technology & Engineering Solutions
of Sandia, LLC (NTESS). Under the terms of Contract DE-NA0003525 with NTESS,
the U.S. Government retains certain rights in this software.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import ghidra.app.decompiler.ClangLine;
import ghidra.app.decompiler.ClangToken;
import ghidra.app.decompiler.ClangTokenGroup;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.decompiler.component.DecompilerUtils;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.FunctionManager;

public class decompile_mapping extends GhidraScript {

    private static final String SCRIPT_NAME = "decompile_mapping.java";

    private void log(String message) {
        println(SCRIPT_NAME + ": " + message);
    }

    private String hexEscape(String input) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (c >= 32 && c <= 126) {
                sb.append(c);
            }
            else {
                sb.append(String.format("\\x%02x", (int) c));
            }
        }
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private void addLineEntry(Map<String, Object> funcMap, String addrKey, String taggedLine, String tokenText) {
        Map<String, Object> entry = (Map<String, Object>) funcMap.get(addrKey);
        if (entry == null) {
            entry = new LinkedHashMap<String, Object>();
            funcMap.put(addrKey, entry);
        }

        List<String> lines = (List<String>) entry.get("line");
        if (lines == null) {
            lines = new ArrayList<String>();
            entry.put("line", lines);
        }
        if (!lines.contains(taggedLine)) {
            lines.add(taggedLine);
        }

        if (tokenText != null) {
            List<String> tokens = (List<String>) entry.get("tokens");
            if (tokens == null) {
                tokens = new ArrayList<String>();
                entry.put("tokens", tokens);
            }
            tokens.add(tokenText);
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> buildMapping() throws Exception {
        log("buildMapping() starting");

        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        FunctionManager functionManager = currentProgram.getFunctionManager();
        int functionCount = functionManager.getFunctionCount();
        log("opened program '" + currentProgram.getName() + "' with " + functionCount + " functions");

        Map<String, Object> outputMap = new LinkedHashMap<String, Object>();
        Map<String, Object> meta = new LinkedHashMap<String, Object>();
        Map<String, Object> functionMeta = new LinkedHashMap<String, Object>();
        outputMap.put("meta", meta);
        meta.put("load_addr", currentProgram.getImageBase().toString());
        meta.put("functions", functionMeta);
        meta.put("function_count", Integer.valueOf(functionCount));

        int processedCount = 0;
        int failedCount = 0;
        List<String> failedFunctions = new ArrayList<String>();

        FunctionIterator functionIter = functionManager.getFunctions(true);
        while (functionIter.hasNext()) {
            Function function = functionIter.next();
            if (function == null) {
                continue;
            }

            String funcName = function.getName(true);
            Map<String, Object> funcMap = (Map<String, Object>) outputMap.get(funcName);
            if (funcMap == null) {
                funcMap = new LinkedHashMap<String, Object>();
                outputMap.put(funcName, funcMap);

                Map<String, Object> funcMeta = new LinkedHashMap<String, Object>();
                funcMeta.put("entry", function.getEntryPoint().toString());
                functionMeta.put(funcName, funcMeta);
            }

            try {
                DecompileResults results = decomp.decompileFunction(function, 120, monitor);
                if (results == null || !results.decompileCompleted()) {
                    log("decompilation did not complete for " + funcName);
                    continue;
                }

                ClangTokenGroup markup = results.getCCodeMarkup();
                if (markup == null) {
                    log("no C markup returned for " + funcName);
                    continue;
                }

                List<ClangLine> lines = DecompilerUtils.toLines(markup);
                for (int i = 0; i < lines.size(); i++) {
                    ClangLine line = lines.get(i);
                    String escapedLine = hexEscape(line.toString());
                    if (escapedLine.length() > 30 && escapedLine.substring(0, 30).contains("WARNING:")) {
                        continue;
                    }

                    String taggedLine = Integer.toString(i + 1) + ": " + escapedLine;
                    boolean lineAdded = false;

                    List<ClangToken> tokens = line.getAllTokens();
                    if (tokens != null) {
                        for (ClangToken token : tokens) {
                            if (token == null) {
                                continue;
                            }

                            Address minAddr = token.getMinAddress();
                            if (minAddr == null) {
                                continue;
                            }

                            addLineEntry(funcMap, minAddr.toString(), taggedLine, hexEscape(token.toString()));
                            lineAdded = true;
                        }
                    }

                    if (!lineAdded) {
                        addLineEntry(funcMap, "None", taggedLine, null);
                    }
                }

                processedCount += 1;
            }
            catch (Exception e) {
                log("ERROR in function " + funcName + ": " + e);
                e.printStackTrace();
                failedFunctions.add(funcName);
                failedCount += 1;
            }
        }

        meta.put("processed_functions", Integer.valueOf(processedCount));
        meta.put("failed_function_count", Integer.valueOf(failedCount));
        if (!failedFunctions.isEmpty()) {
            meta.put("failed_functions", failedFunctions);
        }

        log("buildMapping() finished: processed=" + processedCount + " failed=" + failedCount);
        return outputMap;
    }

    @SuppressWarnings("unchecked")
    private String toJson(Object value) {
        if (value == null) {
            return "null";
        }
        if (value instanceof String) {
            return "\"" + escapeJson((String) value) + "\"";
        }
        if (value instanceof Number || value instanceof Boolean) {
            return value.toString();
        }
        if (value instanceof Map<?, ?>) {
            StringBuilder sb = new StringBuilder();
            sb.append("{");
            boolean first = true;
            for (Map.Entry<Object, Object> entry : ((Map<Object, Object>) value).entrySet()) {
                if (!first) {
                    sb.append(",");
                }
                first = false;
                sb.append(toJson(entry.getKey().toString()));
                sb.append(":");
                sb.append(toJson(entry.getValue()));
            }
            sb.append("}");
            return sb.toString();
        }
        if (value instanceof List<?>) {
            StringBuilder sb = new StringBuilder();
            sb.append("[");
            boolean first = true;
            for (Object item : (List<Object>) value) {
                if (!first) {
                    sb.append(",");
                }
                first = false;
                sb.append(toJson(item));
            }
            sb.append("]");
            return sb.toString();
        }
        return "\"" + escapeJson(value.toString()) + "\"";
    }

    private String escapeJson(String value) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            switch (c) {
                case '\\':
                    sb.append("\\\\");
                    break;
                case '"':
                    sb.append("\\\"");
                    break;
                case '\b':
                    sb.append("\\b");
                    break;
                case '\f':
                    sb.append("\\f");
                    break;
                case '\n':
                    sb.append("\\n");
                    break;
                case '\r':
                    sb.append("\\r");
                    break;
                case '\t':
                    sb.append("\\t");
                    break;
                default:
                    if (c < 32) {
                        sb.append(String.format("\\u%04x", (int) c));
                    }
                    else {
                        sb.append(c);
                    }
            }
        }
        return sb.toString();
    }

    private void writeJson(Path outputPath, Map<String, Object> payload) throws Exception {
        Path parent = outputPath.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }
        Files.write(outputPath, toJson(payload).getBytes(StandardCharsets.UTF_8));
    }

    public void run() throws Exception {
        log("script started");
        String[] args = getScriptArgs();
        log("args = " + java.util.Arrays.toString(args));

        if (args == null || args.length == 0) {
            throw new RuntimeException("No output path argument provided");
        }

        Path outputPath = Paths.get(args[0]);
        Map<String, Object> payload;
        try {
            payload = buildMapping();
        }
        catch (Exception e) {
            log("ERROR while building mapping: " + e);
            e.printStackTrace();
            payload = new LinkedHashMap<String, Object>();
            Map<String, Object> meta = new LinkedHashMap<String, Object>();
            meta.put("error", e.toString());
            meta.put("exception_type", e.getClass().getName());
            payload.put("meta", meta);
        }

        log("writing JSON to " + outputPath);
        writeJson(outputPath, payload);
        log("successfully wrote JSON output");
    }
}
